bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/wrappers/veneer/TestDataClientVeneerApi.java

```diff
@@ -204,30 +234,153 @@ public void testReadRows() throws IOException {
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+
+    ResultScanner resultScanner = dataClientWrapper.readRows(query.createPaginator(100), -1);
+    assertResult(Result.EMPTY_RESULT, resultScanner.next());
+    assertResult(EXPECTED_RESULT, resultScanner.next());
+    assertResult(EXPECTED_RESULT, resultScanner.next());
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                ((ResponseObserver) invocation.getArgument(1)).onComplete();
+                return null;
+              }
+            })
+        .when(mockStreamingCallable)
+        .call(
+            Mockito.any(com.google.cloud.bigtable.data.v2.models.Query.class),
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+
+    resultScanner.close();
+
+    ResultScanner noRowsResultScanner = dataClientWrapper.readRows(query.createPaginator(100), -1);
+    assertNull(noRowsResultScanner.next());
+
+    verify(mockDataClient, times(2)).readRowsCallable(Mockito.<RowResultAdapter>any());
+    verify(mockStreamingCallable, times(2))
+        .call(
+            Mockito.any(Query.class),
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+  }
+
+  @Test
+  public void testReadRowsLowMemory() throws IOException {
+    Query query = Query.create(TABLE_ID).rowKey(ROW_KEY);
+    when(mockDataClient.readRowsCallable(Mockito.<RowResultAdapter>any()))
+        .thenReturn(mockStreamingCallable)
+        .thenReturn(mockStreamingCallable);
+
+    StreamController mockController = Mockito.mock(StreamController.class);
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                cancelled.set(true);
+                return null;
+              }
+            })
+        .when(mockController)
+        .cancel();
+
+    doAnswer(
+            new Answer<Void>() {
+              @Override
+              public Void answer(InvocationOnMock invocation) throws Throwable {
+                ((ResponseObserver) invocation.getArgument(1)).onStart(mockController);
+
+                ((ResponseObserver) invocation.getArgument(1)).onResponse(Result.EMPTY_RESULT);
+                ((ResponseObserver) invocation.getArgument(1)).onResponse(EXPECTED_RESULT);
+                ((ResponseObserver) invocation.getArgument(1)).onComplete();
+                return null;
+              }
+            })
+        .when(mockStreamingCallable)
+        .call(
+            Mockito.any(com.google.cloud.bigtable.data.v2.models.Query.class),
+            Mockito.any(ResponseObserver.class),
+            Mockito.any(GrpcCallContext.class));
+
+    ResultScanner resultScanner = dataClientWrapper.readRows(query.createPaginator(100), 3);
```

Comment:
same question here, why do we return an empty result here? I don't think that's gonna use any memory? Maybe you want to return another expected result and assert readRowsCallable is called twice.

Reply:
Similar to above, I don't think this actually tested the memory buffer function. We should at least return another result, and verify that readRowsCallable are called twice (first time the request got cancelled becuase we filled up the buffer, and the second time we finish the read).